src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
+        if (size == 0) {
+            addHead(newNode);
+        } else if (index == size) {
+            addTail(newNode);
+        } else if (index == 0) {
+            newNode.next = head;
+            head.previous = newNode;
+            head = newNode;
+        } else {
+            Node<T> nodeByIndex = nodeSearch(index);
+            Node<T> previousNode = nodeByIndex.previous;
+            newNode.next = nodeByIndex;
+            nodeByIndex.previous = newNode;
+            newNode.previous = previousNode;
+            previousNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkByIndexEqualSize(index);
+        return nodeSearch(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkByIndexEqualSize(index);
+        Node<T> nodeSearch = nodeSearch(index);
+        T oldValue = nodeSearch.value;
+        nodeSearch.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkByIndexEqualSize(index);
+        Node<T> deletedNode = nodeSearch(index);
+        if (index == 0) {
+            head = head.next;
+        } else if (index == size - 1) {
+            tail = tail.previous;
+        } else {
+            deletedNode.previous.next = deletedNode.next;
+            deletedNode.next.previous = deletedNode.previous;
+        }
+        size--;
+        return deletedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (currentNode.value == object || (currentNode.value != null
+                    && currentNode.value.equals(object))) {
+                if (currentNode == head) {
+                    head = head.next;
+                } else if (currentNode == tail) {
+                    tail = tail.previous;
+                } else {
+                    currentNode.previous.next = currentNode.next;
+                    currentNode.next.previous = currentNode.previous;
+                }
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkByIndexEqualSize(int index) {
+        if (index >= size || index < 0) {
+            trowException();
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index > size || index < 0) {
+            trowException();
+        }
+    }
+
+    private void trowException() {
```

Comment:
this is a bad practice to create a separate method that only throwing an exception